src/dotnet-interactive-vscode/src/fileFormats/interactive.ts

```diff
@@ -0,0 +1,113 @@
+
+const languageSpecifier = '#!';
+
+export function parseAsInteractiveNotebook(contents: string): NotebookDocument {
+    let cells: Array<NotebookCell> = [];
+    let currentLanguage = 'dotnet-interactive.csharp';
+    let lines: Array<string> = [];
+
+    function createCell(language: string, lines: Array<string>): NotebookCell {
+        return {
+            cellKind: language === 'dotnet-interactive.markdown' ? CellKind.Markdown : CellKind.Code,
+            document: {
+                uri: {
+                    fsPath: 'unused'
+                },
+                getText: () => lines.join('\n')
+            },
+            language,
+            outputs: []
+        };
+    }
+
+    function addCell() {
+        // trim from the front
+        while (lines.length > 0 && lines[0] === '') {
+            lines.shift();
+        }
+
+        // trim from the back
+        while (lines.length > 0 && lines[lines.length - 1] === '') {
+            lines.pop();
+        }
+
+        if (lines.length > 0) {
+            cells.push(createCell(currentLanguage, lines));
+        }
+    }
+
+    function addLine(line: string) {
+        lines.push(trimTrailingCarriageReturn(line));
+    }
+
+    for (let line of contents.split('\n')) {
+        if (line.startsWith(languageSpecifier)) {
+            let rawLanguage = line.substr(languageSpecifier.length).trim();
+            let language = editorLanguageAliases.get(rawLanguage);
+            if (language) {
+                // recognized language, finalize the current cell
+                if (lines.length > 0) {
+                    addCell();
+                }
+
+                // found a new cell
+                currentLanguage = language;
+                lines = [];
+            } else {
+                // unrecognized language, probably a magic command
+                addLine(line);
+            }
+        } else {
+            addLine(line);
+        }
+    }
+
+    if (lines.length > 0) {
+        addCell();
+    }
+
+    if (cells.length === 0) {
+        // ensure there's at least one cell available
+        cells.push(createCell('dotnet-interactive.csharp', ['']));
+    }
+
+    return {
+        cells
+    };
+}
+
+export function serializeAsInteractiveNotebook(document: NotebookDocument): string {
+    let lines: Array<string> = [];
+    for (let cell of document.cells) {
+        const cellContents = cell.document.getText();
+        const cellLines = splitAndCleanLines(cellContents);
+        let firstNonBlank = cellLines.findIndex(line => line.length > 0);
+        let lastNonBlank = findIndexReverse(cellLines, line => line.length > 0);
+        if (firstNonBlank >= 0 && lastNonBlank >= 0) {
+            lines.push(`#!${getSimpleLanguage(cell.language)}`);
+            lines.push('');
+            lines.push(...cellLines.slice(firstNonBlank, lastNonBlank + 1));
+            lines.push('');
+        }
+    }
+
+    return lines.join('\r\n');
```

Comment:
Is this potentially changing line endings? Also, is this going to run on Linux and macOS?

Reply:
> Is this potentially changing line endings?

Yes.

> is this going to run on Linux and macOS?

Yes.  This is what we currently do, so there'll be no change with what's currently there.

I considered adding something like `const lineSeparator = isWindows() ? '\r\n' : '\n';`, but my next PR will be to move file serialization/parsing to the kernel and that adds another wrinkle to the mix: it's possible that we may be running the kernel on a different architecture than where the notebook is open, so it's probably best for the notebook to request that the kernel serialize the notebook in a given format with the given line ending.

Reply:
Do we see the user's line ending preferences from VS Code?

Reply:
Added a commit to read the user setting which is either `\n`, `\r\n`, or `auto` (or `undefined` if they don't have the setting).  If it's not a known value, we check if they're on a Windows machine or not to determine.

The helper function added is guaranteed to _only_ return either `\n` or `\r\n`.